File: src/Library/DungeonParams.cpp
/*
* @author PELLETIER Benoit
*
* @file DungeonParams.cpp
*
* @date 23/10/2018
*
* @brief Define all params of a dungeon generation
*
*/

#include "DG/DungeonParams.h"

const DungeonParams DungeonParams::zero(
	0, // width
	0, // height
	0, // roomMinWidth
	0, // roomMinHeight
	0, // roomMaxWidth
	0, // roomMaxHeight
	0, // iteration
	0, // maxChest
	0, // maxChestPerRoom
	0, // maxEnemy
	0  // maxEnemyPerRoom
	);

const DungeonParams DungeonParams::basic(
	64, // width
	64, // height
	3, // roomMinWidth
	3, // roomMinHeight
	0, // roomMaxWidth
	0, // roomMaxHeight
	5, // iteration
	5, // maxChest
	2, // maxChestPerRoom
	20,// maxEnemy
	5  // maxEnemyPerRoom
	);

DungeonParams::DungeonParams(uint64_t _width, uint64_t _height, uint64_t _roomMinWidth, uint64_t _roomMinHeight, uint64_t _roomMaxWidth, uint64_t _roomMaxHeight, uint8_t _iteration, uint8_t _maxChest, uint8_t _maxChestPerRoom, uint8_t _maxEnemy, uint8_t _maxEnemyPerRoom)
	: width(_width), height(_height), roomMinWidth(_roomMinWidth), roomMinHeight(_roomMinHeight), roomMaxWidth(_roomMaxWidth), roomMaxHeight(_roomMaxHeight), iteration(_iteration), maxChest(_maxChest), maxChestPerRoom(_maxChestPerRoom), maxEnemy(_maxEnemy), maxEnemyPerRoom(_maxEnemyPerRoom)
{
}

DungeonParams Lib_DungeonParams_constructor(uint64_t _width, uint64_t _height, uint64_t _roomMinWidth, uint64_t _roomMinHeight, uint64_t _roomMaxWidth, uint64_t _roomMaxHeight, uint8_t _iteration, uint8_t _maxChest, uint8_t _maxChestPerRoom, uint8_t _maxEnemy, uint8_t _maxEnemyPerRoom)
{
	return DungeonParams(_width, _height, _roomMinWidth, _roomMinHeight, _roomMaxWidth, _roomMaxHeight, _iteration, _maxChest, _maxChestPerRoom, _maxEnemy, _maxEnemyPerRoom);
}

File: UnitySample/Assets/Plugins/DungeonGenerator/DungeonParams.cs
using System;
using System.Runtime.InteropServices;

[StructLayout(LayoutKind.Explicit, CharSet = CharSet.Unicode)]
public struct DungeonParams
{
    [FieldOffset(0)] public ulong width;
    [FieldOffset(8)] public ulong height;
    [FieldOffset(16)] public ulong roomMinWidth;
    [FieldOffset(24)] public ulong roomMinHeight;
    [FieldOffset(32)] public ulong roomMaxWidth;
    [FieldOffset(40)] public ulong roomMaxHeight;
    [FieldOffset(48)] public char iteration;
    [FieldOffset(49)] public char maxChest;
    [FieldOffset(50)] public char maxChestPerRoom;
    [FieldOffset(51)] public char maxEnemy;
    [FieldOffset(52)] public char maxEnemyPerRoom;
    [FieldOffset(53)] public char nbLockedRoom;
    [FieldOffset(54)] public bool randomSeed;
    [FieldOffset(56)] public ulong seed;

    static DungeonParams zero = new DungeonParams {
        width = 0,
        height = 0,
        roomMinWidth = 0,
        roomMinHeight = 0,
        roomMaxWidth = 0,
        roomMaxHeight = 0,
        iteration = (char)0,
        maxChest = (char)0,
        maxChestPerRoom = (char)0,
        maxEnemy = (char)0,
        maxEnemyPerRoom = (char)0,
        nbLockedRoom = (char)0,
        seed = 0,
        randomSeed = false
    };

    static DungeonParams basic = new DungeonParams
    {
        width = 64,
        height = 64,
        roomMinWidth = 3,
        roomMinHeight = 3,
        roomMaxWidth = 0,
        roomMaxHeight = 0,
        iteration = (char)5,
        maxChest = (char)5,
        maxChestPerRoom = (char)2,
        maxEnemy = (char)20,
        maxEnemyPerRoom = (char)5,
        nbLockedRoom = (char)1,
        seed = 0,
        randomSeed = true
    };
}
